torchft/process_group.py

```diff
@@ -716,23 +750,76 @@ def _worker(
                 return
             tx.put(None)
 
-            work = {}
+            streams: Dict[str, torch.cuda.Stream] = {}
+            work: Dict[int, _OpMetadata] = {}
             next_op_id: int = 0
 
             while True:
                 op = rx.get()
                 cmd = op[0]
                 if cmd == "func":
-                    func_name, args, kwargs = op[1:]
-                    args = _PickleSafeOptions.unsafe_args(args)
-                    fn = getattr(pg, func_name)
-                    work[next_op_id] = fn(*args, **kwargs)
+                    func_name, args, kwargs, stream_device, stream_id, event = op[1:]
+
+                    print(f"func {func_name=}")
+
+                    # To avoid potential deadlocks we need to preserve the
+                    # stream/synchronization behavior of the parent process.
+                    # We allocate one Stream per stream_id to make sure that we
+                    # don't accidentally introduce cross stream synchronization
+                    # points.
+                    if stream_id is not None:
+                        stream_key = f"{stream_device}/{stream_id}"
+                        if stream_key not in streams:
+                            streams[stream_key] = torch.cuda.Stream(
+                                device=stream_device
+                            )
```

Comment:
Are we going to have zombie `stream` if there are multiple failures? Will this cause memory leakage?

Reply:
My understanding is streams are specific to the cuda context/process so this will be cleaned up just fine when it gets killed